### PR TITLE
Enable annotations mapping to be optional

### LINKED ISSUE #1088: Binding to an Opstree Solutions Redis instance with no redisSecret fails
### What is the environment (Minikube, Openshift)?

CodeReady Containers version: 1.37.0+3876d27d
OpenShift version: 4.9.10 (embedded in executable)

### What is the SBO version used?

v1.0.0

### What are the steps to reproduce this issue?

  1. Create a redis operator with the following specs:

```
apiVersion: redis.redis.opstreelabs.in/v1beta1
kind: Redis
metadata:
  name: redis1
spec:
  kubernetesConfig:
    image: quay.io/opstree/redis:v6.2.5
    imagePullPolicy: IfNotPresent
    resources:
      limits:
        cpu: 101m
        memory: 128Mi
      requests:
        cpu: 101m
        memory: 128Mi
    serviceType: ClusterIP
  redisExporter:
    enabled: false
    image: quay.io/opstree/redis-exporter:1.0
  storage:
    volumeClaimTemplate:
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 1Gi
```

  2. Create a link with the specs:

```
apiVersion: binding.operators.coreos.com/v1alpha1
kind: ServiceBinding
metadata:
  name: redis-binding
spec:
  application:
    group: apps
    name: my-deployment
    resource: deployments
    version: v1
  bindAsFiles: false
  detectBindingResources: true
  services:
  - group: redis.redis.opstreelabs.in
    id: redis1
    kind: Redis
    name: redis1
    version: v1beta1
```

### What is the expected behaviour?

The service binding operator should create the binding

### What is the actual behaviour?

The creation of the binding fails due to a not found `redisSecret` field.

The `redisSecret` is not mandatory in the `redis` CRD.

```
status:
  conditions:
  - lastTransitionTime: "2022-01-10T13:55:02Z"
    message: redisSecret is not found
    reason: ErrorReadingBinding
    status: "False"
    type: CollectionReady
  - lastTransitionTime: "2022-01-10T13:55:02Z"
    message: redisSecret is not found
    reason: ProcessingError
    status: "False"
    type: Ready
  secret: ""
```

```
$ kubectl get crd redis.redis.redis.opstreelabs.in -o yaml 
[...]
    service.binding/host: path={.metadata.name}
    service.binding/password: path={.spec.kubernetesConfig.redisSecret.name},objectType=Secret,sourceKey=password
    service.binding/type: redis
[...]
```

### Service Binding Operator Logs

```
{"level":"info","ts":1641823073.7587097,"logger":"controllers.ServiceBinding","msg":"Reconciling","serviceBinding":"prj1/nodejs-x5314-oasw-redis-redis1"}
{"level":"info","ts":1641823073.8065016,"logger":"controllers.ServiceBinding","msg":"Done","serviceBinding":"prj1/nodejs-x5314-oasw-redis-redis1","retry":true,"error":"redisSecret is not found"}
{"level":"error","ts":1641823073.806546,"logger":"controller","msg":"Reconciler error","reconcilerGroup":"binding.operators.coreos.com","reconcilerKind":"ServiceBinding","controller":"servicebinding","name":"nodejs-x5314-oasw-redis-redis1","namespace":"prj1","error":"redisSecret is not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/tmp/github.com/redhat-developer/service-binding-operator/vendor/github.com/go-logr/zapr/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/tmp/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:246\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/tmp/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/tmp/github.com/redhat-developer/service-binding-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:197\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/tmp/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/tmp/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/tmp/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/tmp/github.com/redhat-developer/service-binding-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90"}
```

## Release Notes
Currently, binding to a `Redis` CR from the `redis.redis.opstreelabs.in/v1beta1` API fails due to the absence of an optional `redisSecret` field in the `Redis` CR. There is currently no workaround for this issue.

### PR DESCRIPTION
**Pull request by baijum**
Fixes #1088

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/redhat-developer/service-binding-operator/blob/master/CONTRIBUTING.md#docs) 
  included if any changes are user facing
- [x] [Tests](https://github.com/redhat-developer/service-binding-operator/blob/master/CONTRIBUTING.md#tests)
  included if any functionality added or changed. For bugfixes please include tests that can catch regressions
- [x] All acceptance test scenarios included in the PR which verifies a bugfix or a requested feature reported by a non-member are tagged with `@external-feedback` tag.
- [ ] Follows the [commit message standard](https://github.com/redhat-developer/service-binding-operator/blob/master/CONTRIBUTING.md#commits)

### CODE REVIEW

**Comment by feloy** [comment]
Do we need to add the `optional=true` option to the redis CRD annotation manually, or will it be added by the service binding operator?

**Review by sadlerap — CHANGES_REQUESTED** [review]
Mostly looks good to me, a few changes though:
- There are some warnings from the linter that need to be addressed.
- I would like to see some unit tests added.

**Comment by baijum** [comment]
> Do we need to add the `optional=true` option to the redis CRD annotation manually, or will it be added by the service binding operator?

I have made that change in this PR.

**Comment by feloy** [comment]
This PR will fix the original issue.  Thanks for this work
LGTM

**Comment by baijum** [comment]
> * There are some warnings from the linter that need to be addressed.

Fixed.

> * I would like to see some unit tests added.

Done.

File: test/acceptance/features/bindAppToServiceAnnotationsOptional.feature [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            """
+        Then Service Binding is ready
+        * File "/bindings/$scenario_id/uri" is unavailable in application pod
+        * Content of file "/bindings/$scenario_id/image" in application pod is
+            """
+            busybox
+            """
+
+    @negative
+    Scenario: Bind to the application with annotations and with missing non-optional source field to cause binding failure
```
Comment (baijum) [inline test/acceptance/features/bindAppToServiceAnnotationsOptional.feature:100]: Add a scenario with `optional=false` (feedback from commmunity call)

Comment (baijum) [reply test/acceptance/features/bindAppToServiceAnnotationsOptional.feature:100]: Added the test case.

**Review by sadlerap — APPROVED** [review]
/lgtm

@Srivaralakshmi should probably take a look at the docs changes.

**Review by Srivaralakshmi — CHANGES_REQUESTED** [review]
@baijum Left my suggestions. PTAL.Thanks!

File: docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc [hunk 2]
```diff
@@ -294,3 +294,21 @@ The following example shows how the previous values of a collection in annotatio
 ----
 
 The previous example helps you project all those values with key such as `prefix_0`, `prefix_1`, and so on. The default prefix is the name of the resource `kind:`.
+
+[#setting-annotations-mapping-to-be-optional] 
+== Setting annotations mapping to be optional
+
+Sometimes a field may not exist in the target path. By default, it will create an error. If these is a field that may not exist, you can add the `optional=true` flag as part of the annotation value.
```
[the remark this answers is not in the document]
Comment (baijum) [reply docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc]: @Srivaralakshmi  I can see the whole paragraph has been rewritten. I am not sure which sentence that you want me to rewrite.

[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc]: @baijum I meant this line: If these is a field that may not exist, you can add the `optional=true` flag as part of the annotation value.

But the request was only if my suggestion seems inaccurate technically.

File: docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc [hunk 3]
```diff
@@ -294,3 +294,33 @@ The following example shows how the previous values of a collection in annotatio
...
+[#setting-annotations-mapping-to-be-optional] 
+== Setting annotations mapping to be optional
+
+You can have optional fields in the annotations. For example, a link to the credentials might not be there if the service endpoint does not require authentication. In such cases, a field might not exist in the target path of the annotations. As a result, the {servicebinding-title} generates an error, by default. As a service provider, to indicate whether you require annotations mapping, you can set a value to the `optional` flag in your annotations while enabling services.
+
+{servicebinding-title} provides annotations mapping only if the target path is available. When the target path is not available the {servicebinding-title} skips providing the annotations mapping. 
+
+.Procedure
+
+. To enable {servicebinding-title} consider a new field in the annotations, provide the mapping by setting the `optional` flag value to `true`.
```
Comment (Srivaralakshmi) [inline docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc:307]:
```suggestion
. To enable {servicebinding-title} consider a new field in the annotations, provide the mapping by setting the `optional` flag value to `true`:
```

File: docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc [hunk 4]
```diff
@@ -294,3 +294,33 @@ The following example shows how the previous values of a collection in annotatio
...
+  name: my-db
+  namespace: my-petclinic
+  annotations:
+    service.binding/username: path={.spec.name},optional=true
+    ...
+----
+
+[NOTE]
+====
+* If you set the `optional` flag value to `false` and when the {servicebinding-title} is unable to find the target path, the annotations mapping fails.
```
Comment (Srivaralakshmi) [inline docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc:324]:
```suggestion
* If you set the `optional` flag value to `false` and when the {servicebinding-title} is unable to find the target path, the Operator fails the annotations mapping.
```

**Review by Srivaralakshmi — CHANGES_REQUESTED** [review]
@baijum Left my suggestions. PTAL.Thanks!

File: docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (baijum) [reply docs/userguide/modules/exposing-binding-data/pages/adding-annotation.adoc]: The current text looks good to me. Please approve the PR if no more issues.

**Review by Srivaralakshmi — COMMENTED** [review]
@baijum Approved!

**Review by Srivaralakshmi — APPROVED** [review]
@baijum Approved!

**Pull request merged**